### PR TITLE
Draft: Adapt to s3 0.0.6

### LINKED ISSUE #3: Replace list of attachmentPaths by list of attachments.
**Is your feature request related to a problem? Please describe.**
It can be possible that an email attaches file from multiple s3-services. As a result, the url for each s3-service is needed.

**Describe the solution you'd like**

Replace in class `io.muenchendigital.digiwf.email.integration.domain.model.Mail` list of attachementPath of type string by list of attachments of type "Attachemt".

The Class attachment contains two attributes.

```java
public class Attachment {

    private String documentStorageUrl;

    private String pathToFile;

}
```

These two attributes can be used with the client lib.

```java
        final byte[] binaryFile = this.documentStorageFileRepository.getFile(
                attachment.getPathToFile(),
                attachment.getDocumentStorageUrl,
                3 // Expiration time
        );
```

**Describe alternatives you've considered**
There is no alternative.

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by axd4**
---
name: Pull request
about: Contribute your changes to the project
title: '[pull-request]'
labels: ''
assignees: ''
---
**Description**
Short description or comments

**Reference**
Issues #XXX

**Checklist**

- [ ] Acceptance criteria are met
- [ ] Pipeline has been run successfully
- [ ] JUnit tests have been run successfully
- [ ] Changelog is adjusted
- [ ] [Documentations](https://git.muenchen.de/groups/digitalisierung/-/wikis/Dokumentationen) are completed
- [ ] Frontend is tested
- [ ] Created sub-branches are deleted
- [ ] Boards are updated ( [Digitalisierung](https://git.muenchen.de/groups/digitalisierung/-/boards) , [Support](https://git.muenchen.de/digitalisierung/digiwf-support) , [Zenhub](https://app.zenhub.com/workspaces/digiwf-621f70bf50ea1100120b7e93/board) )
- [ ] feature process is created or extended
- [ ] [git-ops](https://git.muenchen.de/digitalisierung/digiwf-ops) is customized
- [ ] Openshift environments are prepared (Secrets, etc.)

### CODE REVIEW

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+     * Path to the file inside in the S3 storage.
+     */
+    private String attachmentPath;
+
+    /**
+     * Optional filename. If not set, the filename of the file in the S3 Storage is used.
+     */
+    private String fileName;
+
+    public boolean hasRequiredData() {
```
Comment (boal) [inline digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java:31]: would it be better if the method name is "areAttributesNotBlank"?

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        return StringUtils.isNotBlank(attachmentPath) && StringUtils.isNotBlank(documentStorageUrl);
+    }
+
+    public boolean hasFileName() {
```
Comment (boal) [inline digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java]: Same here. Suggestion: "isFilenameNotBlank"

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Mail.java [hunk 3]
```diff
@@ -45,14 +45,14 @@ public class Mail {
      */
     private String replyTo;
 
-    private List<String> attachmentPaths;
+    private List<Attachment> attachments;
 
     public boolean hasAttachement() {
```
Comment (boal) [inline digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Mail.java]: Rename hasAttachements. Current name is singular

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Mail.java [hunk 4]
```diff
@@ -45,14 +45,14 @@ public class Mail {
...
-        return this.attachmentPaths != null && this.attachmentPaths.size() > 0;
+        return this.attachments != null && this.attachments.size() > 0;
     }
 
     public boolean hasReplyTo() {
```
Comment (boal) [inline digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Mail.java]: Same here: "isReplyToNotBlank"

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import lombok.RequiredArgsConstructor;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Attachment File you want to get from the S3 storage.
+ */
+@Data
+@RequiredArgsConstructor
+@AllArgsConstructor
+public class Attachment {
```
Comment (boal) [inline digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java:13]: General comment: Attribute validation should not be included within the model/dto/entity class. Better is to write in own validation class.

File: example-digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/api/controller/ExampleController.java [hunk 6]
```diff
@@ -47,9 +47,13 @@ private Mail getMail() {
         mail.setSubject("Test1234");
         mail.setBody("Hallo test123");
         mail.setReplyTo("");
-//        final List<String> attachmentPaths = new ArrayList<>();
-//        attachmentPaths.add("test/picture.jpg");
-//        mail.setAttachmentPaths(attachmentPaths);
+//        Attachment
```
Comment (boal) [inline example-digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/api/controller/ExampleController.java]: remove commented out code

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java (hunk 1 above)
Comment (axd4) [reply digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java:31]: Could do, but it's not all attributes that get checked here, just the required ones. Maybe "hasRequiredAttributesFilled"? Or "areRequiredAttributesFilled"?

Comment (boal) [inline digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java]: would it be better if the method name is "areAttributesNotBlank"?

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java (hunk 2 above)
Comment (axd4) [reply digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java]: Removed method

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Mail.java (hunk 3 above)
Comment (axd4) [reply digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Mail.java]: Removed method

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Mail.java (hunk 4 above)
Comment (axd4) [reply digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Mail.java]: Removed method

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java (hunk 5 above)
Comment (axd4) [reply digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java:13]: Removed methods

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java (hunk 1 above)
Comment (axd4) [reply digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/model/Attachment.java]: Removed method

File: example-digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/api/controller/ExampleController.java (hunk 6 above)
Comment (axd4) [reply example-digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/api/controller/ExampleController.java]: Moved to own method, only left one commented out code in with a comment to comment it in if needed.

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/service/MailingService.java [hunk 7]
```diff
@@ -62,20 +63,27 @@ public void sendMail(final Mail mail) throws MissingInformationMailException {
...
-                for (val attachmentPath : mail.getAttachmentPaths()) {
-                    final byte[] binaryFile = this.documentStorageFileRepository.getFile(
-                            attachmentPath,
-                            3
-                    );
-                    final Tika tika = new Tika();
-                    val file = new ByteArrayDataSource(binaryFile, tika.detect(binaryFile));
-                    val fileName = StringUtils.substringAfterLast(attachmentPath, "/");
-                    helper.addAttachment(fileName, file);
+            if (mail.getAttachments() != null && mail.getAttachments().size() > 0) {
```
Comment (boal) [inline digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/service/MailingService.java]: put validation logic in a method.

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/service/MailingService.java [hunk 8]
```diff
@@ -62,20 +63,27 @@ public void sendMail(final Mail mail) throws MissingInformationMailException {
...
+                for (val attachment : mail.getAttachments()) {
+                    if (StringUtils.isNotBlank(attachment.getAttachmentPath()) && StringUtils.isNotBlank(attachment.getDocumentStorageUrl())) {
```
Comment (boal) [inline digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/service/MailingService.java]: put validationlogic in a method.

e.g. "isAttachmentPathAndDocumentStorageNotBlank"

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/service/MailingService.java (hunk 7 above)
Comment (axd4) [reply digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/service/MailingService.java]: done

File: digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/service/MailingService.java (hunk 8 above)
Comment (axd4) [reply digiwf-email-integration/src/main/java/io/muenchendigital/digiwf/email/integration/domain/service/MailingService.java]: done

**Review by boal — APPROVED** [review]

**Review by boal — APPROVED** [review]

**Pull request merged**